Examples/ThirdParty/Spotify/Sources/main.swift

```diff
@@ -0,0 +1,231 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// NOTE: This file is automatically-generated!
+import Foundation
+import Dispatch
+import OAuth2
+import GoogleAPIRuntime
+import Commander
+
+let CLIENT_CREDENTIALS = "spotify.json"
+let TOKEN = "spotify.json"
+
+func main() throws {
+  let scopes = [
+    "user-modify-playback-state",
+    "user-read-birthdate",
+    "user-read-currently-playing",
+    "user-read-email",
+    "user-read-playback-state",
+    "user-read-private"]
+
+  guard let tokenProvider = BrowserTokenProvider(credentials:CLIENT_CREDENTIALS, token:TOKEN) else {
+    return
+  }
+  let spotify = try Spotify(tokenProvider:tokenProvider)
+
+  let group = Group {
+    $0.command("login", description:"Log in with browser-based authentication.") {
+      try tokenProvider.signIn(scopes:scopes)
+      try tokenProvider.saveToken(TOKEN)
+    }
+
+    $0.command(
+      "artists.get",
+      Options<String>("p_id", default: [], count: 1, description: "The Spotify ID of the artist."),
```

Comment:
where does the `p_` come from? I don't see `p_id`, just `id` in the schema. Just want to understand the decision to prefix the parameters. 😄

Reply:
It's pretty awful. The Discovery Format allows API calls to have parameters and request objects. It's possible in rare cases for parameters to have the same names as fields in the request objects. I added these prefixes to get the generated Compute API to compile but would like to come up with something better.

Reply:
ah yikes thats a tricky one. so `p_` is for parameter and `r_` is for request body? Hm. While that is concise, its not immediately obvious imo. And this detail of the schema is bleeding into the UX of the generated CLI. As a user I'd ask "If I have some data, how do I know where to put it?"

Would `param_id` or `param.id` / `body_id` or `body.id` be easier to understand? Just throwing something out there. No need to decide here.

Reply:
I had been thinking that it would be good to just follow [google-api-rs](https://github.com/Byron/google-apis-rs) on this, but it too is lumping request and parameter fields into separate categories (see the [Language v1 CLI docs](http://byron.github.io/google-apis-rs/google_language1_cli/index.html)). It also isn't providing much help information (compared to [protoc-gen-go_cli](https://github.com/googleapis/gapic-generator-go/tree/master/cmd/protoc-gen-go_cli)). 

In some cases, the overlap can be resolved because the request object fields are marked "output only". But unfortunately, the only indicator of that is "[Output Only]" at the beginning of their description string, and even after accounting for that, I still found collisions with the Compute API.

Reply:
Another problem here is that the Commander library that I'm using to build the CLI has a [hard-coded maximum number of command options](https://github.com/kylef/Commander/blob/314f8d7455e25e60e9e9b850040eebd0a87f2527/Sources/Commander/Commands.swift#L538) that we overrun when we generate a CLI for Compute.